metricbeat/module/linux/perf/perf.go

```diff
@@ -0,0 +1,277 @@
+		err := startMonitor(procList)
+		if err != nil {
+			return nil, errors.Wrap(err, "error starting monitor")
+		}
+	}
+
+	// This perf library is Not So Good and eats errors that come from perf_event_open
+	// Do a quck test using a lower-level API to make sure we can actually access perf APIs
+	testP, err := perf.NewMinorFaultsProfiler(procList[0].PID, -1)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating profiler for Minor Page faults")
+	}
+	testP.Close()
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		processes:     procList,
+		period:        config.SamplePeriod,
+		continuous:    isContinuoius,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	perfData, err := runSampleForPeriod(m.processes, m.period, m.continuous)
+	if err != nil {
+		return errors.Wrap(err, "error running sample")
+	}
+
+	for _, process := range perfData {
+
+		event := common.MapStr{}
+
+		if process.SwMetrics.TimeRunning != nil {
+			event["software_events"] = process.SwMetrics
+		}
+
+		if process.HwMetrics.TimeRunning != nil {
+			event["hardware_events"] = process.HwMetrics
+		}
+
+		report.Event(mb.Event{
+			MetricSetFields: event,
+			RootFields: common.MapStr{
+				"process": common.MapStr{
+					"pid":  process.Metadata["pid"],
+					"ppid": process.Metadata["ppid"],
+					"pgid": process.Metadata["pgid"],
+					"name": process.Metadata["name"],
+				},
+			},
+		})
+	}
+
+	return nil
+}
+
+// in cases where we're not stopping monitoring, start before we call `runSampleForPeriod`.
+// Otherwise call every sample period.
+func startMonitor(toSample []procInfo) error {
+	for _, pid := range toSample {
+
+		if pid.HardwareProc != nil {
+			err := pid.HardwareProc.Start()
+			if err != nil {
+				return errors.Wrap(err, "error starting HW profiler")
+			}
+		}
+
+		if pid.SoftwareProc != nil {
+			err := pid.SoftwareProc.Start()
+			if err != nil {
+				return errors.Wrap(err, "error starting SW profiler")
+			}
+		}
+
+	}
+	return nil
+}
+
+// runSampleForPeriod starts a sample, sleeps for a given period, then collects metrics.
+func runSampleForPeriod(toSample []procInfo, period time.Duration, isContinuoius bool) ([]perfInfo, error) {
+	if !isContinuoius {
+		err := startMonitor(toSample)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	time.Sleep(period)
+
+	var metrics = []perfInfo{}
+
+	for _, pid := range toSample {
+
+		newMetric := perfInfo{}
+		newMetric.Metadata = pid.Metadata
+
+		if pid.HardwareProc != nil {
+			hwPro, err := pid.HardwareProc.Profile()
+			if err != nil {
+				return nil, errors.Wrap(err, "error gathering HW profile")
+			}
+			newMetric.HwMetrics = *hwPro
+
+			if !isContinuoius {
+				pid.HardwareProc.Stop()
+			}
+			pid.HardwareProc.Reset()
```

Comment:
Consider moving the stop and reset calls into their own function. Then `defer` that function call right after your call `startMonitor` at the start of this function. That way, we'll stop and reset things even if there are errors.

At that point, there's perhaps even an opportunity to call this stop-and-reset function from the metricset's `Close()` function below?

Reply:
Was wondering about that, but I wanted the `Stop` and `Reset` calls to be as 'close' to the `Profile` calls as possible, to minimize any lost counters or weird metrics.

Reply:
Well, with the `defer` the `Stop` and `Reset` calls won't happen until this `runSampleForPeriod` returns so you shouldn't lose any counters.

Reply:
My concern with the way it's implemented at the moment is that if there are any errors between the point when `startMonitor` is called and here, we would've called `pid.HardwareProc.Start()` and `pid.SoftwareProc.Start()` for every PID being monitored and then returned from `runSampleForPeriod` because of an error. We would have never called ``pid.HardwareProc.Stop()` and `pid.SoftwareProc.Stop()`.

I'm not sure if this poses any risks either with leaking resources or performance or collection of bad data. If it doesn't, then what you have here is fine as-is.

Reply:
Yah, I get where you're coming from. Might be best to just err on the side of safety?